subscribe.go

```diff
@@ -168,51 +171,65 @@ func (ls *legSubscriber) watch(ctx context.Context, sub *pubsub.Subscription) {
 			return
 		}
 
-		// Run policy from pubsub message to see if exchange needs to be run
-		allow := true
+		// Run policy from pubsub message to see if exchange needs to be run.
+		//
+		// TODO: Change this to call callback to evaluate peer using policy
+		// callback provided to go-legs.
 		if ls.policy != nil {
+			var allow bool
 			ls.hndmtx.RLock()
 			allow, err = ls.policy(msg)
 			if err != nil {
 				log.Errorf("Error running policy: %v", err)
 			}
 			ls.hndmtx.RUnlock()
+			if !allow {
+				log.Infow("Message from peer not allowed by legs policy")
+				return
+			}
 		}
 
-		if allow {
-			log.Debugf("Pubsub message received. Policy says we are allowed to process")
-			src, err := peer.IDFromBytes(msg.From)
-			if err != nil {
-				continue
-			}
+		log.Debugf("Pubsub message received. Policy says we are allowed to process")
+		src, err := peer.IDFromBytes(msg.From)
+		if err != nil {
+			continue
+		}
 
-			c, err := cid.Cast(msg.Data)
-			if err != nil {
-				log.Warnf("Couldn't cast CID from pubsub message")
-				continue
-			}
-			v := Voucher{&c}
-
-			// Locking latestSync to avoid data races by several updates
-			// This sync is unlocked when the transfer is finished or if
-			// there is an error in the channel.
-			ls.syncmtx.Lock()
-			log.Debugf("Starting data channel (cid: %s, latestSync: %s)", c, ls.latestSync)
-			ls.syncing = c
-			_, err = ls.dt.OpenPullDataChannel(ctx, src, &v, c,
-				ExploreRecursiveWithStopNode(
-					selector.RecursionLimitNone(),
-					ls.dss,
-					ls.latestSync))
-			if err != nil {
-				// Log error for now.
-				log.Errorf("Error in data channel: %v", err)
-				// There has been an error, no FinishTransfer event will be triggered,
-				// so we need to release the lock here.
-				ls.syncmtx.Unlock()
+		// Decode cid and originator addresses from message.
+		c, addrs, err := decodeMessage(msg.Data)
+		if err != nil {
+			log.Warnf("Could not decode pubsub message: %s", err)
+			continue
+		}
+		v := Voucher{&c}
+
+		// Add the message originator's address to the peerstore.  This
+		// allows a connection, back to that provider that sent the
+		// message, to retrieve advertisements.
+		peerStore := ls.host.Peerstore()
+		if peerStore != nil {
+			peerStore.AddAddrs(src, addrs, 5*time.Minute)
```

Comment:
make a constant or var for time

i wonder if we want to keep known addresses for the peer around for the life of a subscriber so that if an explicit sync is called we know we have them?

Reply:
If we do explicit sync, that means that we already know about the provider and that it is registered with the indexer and its address is known. So for manual sync, the provider address could be put into the peerstore at that time.

Reply:
Used the existing const `peerstore.ProviderAddrTTL` which seems all too appropriate.